### PR TITLE
[RFC] Update interrupt framework to support control queue

### PR DESCRIPTION
**Pull request by bjzhjing**
According to virtio spec 1.1, multiple queue support (MQ) for net device requires control queue implemented. Refer to the process in guest kernel, it keeps to assign an invalid msix vector value (0xfffff) to control queue, which means control queue has no valid PCI-MSIX interrupt for use, it should be set with ISR instead. According to this, for virtio net device with control queue enabled, both ISR and PCI-MSIX interrupt mode should be supported. However, cloud-hypervisor does not support such case, which assigns virtio device with either ISR or PCI-MSIX interrupt. Update the interrupt framework on cloud-hypervisor to resolve this issue.

Refer to the following links to see the detail on how guest kernel setup control queue, since callback for control queue is null, it will always get VIRTIO_MSI_NO_VECTOR (0xffff)
vp_find_vqs_msix in 
https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_pci_common.c

In cloud-hypervisor, according to queues.rs, vector of virtqueue is initialized with 0. Thus, in PR #409, control queue takes the index 0 of PCI-MSIX table for use, which is actually assigned to config change from guest kernel side.

The current change is based on the existing interrupt framework with the following items, mark it as RFC for discussion.

- Add new method in VirtioDevice trait to tell if virtio device needs ISR support
- Update device manager to assign interrupts according to virtio devices' requirement
- Update all virtio devices' activate process to handle all interrupt modes
- Add control queue handling module
- Enable control queue support for virtio-net to see if it work normally under new update of interrupt framework above
 
Note:
To address #479 
PR #409 has dependency on this one

### CODE REVIEW

**Comment by bjzhjing** [comment]
@sboeuf @sameo @rbradford 
Hi, please take a look!

**Comment by sboeuf** [comment]
@bjzhjing 
>According to virtio spec 1.0


>multiple queue support (MQ) for net device requires control queue implemented, and control queue will communicate with guest kernel through ISR, which means that for virtio net device with control queue enabled, both ISR and PCI-MSIX interrupt mode should be supported at the same time. However, cloud-hypervisor does not support such case, which will assign virtio device with either ISR or PCI-MSIX interrupt. Update the interrupt framework on cloud-hypervisor to resolve this issue.

**Comment by bjzhjing** [comment]
@sboeuf Sorry for making you misunderstand. I have update the description above, let me know if you still have any concern.

**Comment by sboeuf** [comment]
@bjzhjing 
>Refer to the process in guest kernel, it keeps to assign an invalid msix vector value (0xfffff) to control queue, which means control queue has no valid PCI-MSIX interrupt for use, it should be set with ISR instead.

The spec defines that if the driver sets `0xfffff`, it wants to disable the interrupt: 
```
To disable interrupts for an event type, the driver unmaps this event by writing a special
NO_VECTOR value.
```

Now the question is, why does the driver try to disable the interrupt for the `controlq` queue? Looking into [this](https://github.com/torvalds/linux/blob/master/drivers/net/virtio_net.c#L2798) you can see the function `find_vqs()` is being invoked with a set of callbacks, but for the `controlq`, it is hardcoding the callback to `NULL` (look [here](https://github.com/torvalds/linux/blob/master/drivers/net/virtio_net.c#L2780-L2784)).
And because the callback is `NULL` later in the code [here](https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_pci_common.c#L317-L318), the driver ends up setting the MSI-X vector to the value `VIRTIO_MSI_NO_VECTOR`. This does not mean the function `vp_find_vqs_msix()` will return an error, but simply that the driver is not setting up an IRQ handler for this queue. It's going to poll and that means we should not try to generate an interrupt from our `virtio-net` device.

To summarize, I think that from the code of the interrupt handler we have in our VMM, we should check the value of the vector [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vm-virtio/src/transport/pci_device.rs#L502), and if it appears the value is `0xFFFF`, we should simply return without doing anything.

**Comment by sboeuf** [comment]
@bjzhjing 
Based on my reading of the kernel code, and based on the following experiment, I don't see a reason why we would need INTx to be implemented for the `controlq`.

Here is the experiment:
- This CH branch contains one patch adding the `controlq`: https://github.com/sboeuf/cloud-hypervisor-1/tree/virtio-net-ctl-queue
- This kernel branch enables some logs to check the `virtio-net` driver detects a control queue is present, and that it never falls back onto the INTx initialization: https://github.com/sboeuf/linux/tree/virtio-net-ctl-queue

Please try on your side and tell me what your findings are. But I really believe this current PR is not needed and should be closed.

**Comment by bjzhjing** [comment]
@sboeuf Yes, you're right, there is no INTx initialization for control queue. While guest kernel still need to be notified by VMMs once control queue handling is done. Actually, this is achieved through ISR, but no irq injection required. guest kernel get ISR from the following code path. 

> guest kernel side:

```
int virtio_pci_modern_probe(struct virtio_pci_device *vp_dev)
{
...
        isr = virtio_pci_find_capability(pci_dev, VIRTIO_PCI_CAP_ISR_CFG,
                                         IORESOURCE_IO | IORESOURCE_MEM,
                                         &vp_dev->modern_bars);
...
       vp_dev->isr = map_capability(pci_dev, isr, sizeof(u8), 1,
                                     0, 1,
                                     NULL);
...
}
```

> cloud-hypervisor side:

```
fn read_bar(&mut self, _base: u64, offset: u64, data: &mut [u8]) {
        match offset {
...
            o if ISR_CONFIG_BAR_OFFSET <= o && o < ISR_CONFIG_BAR_OFFSET + ISR_CONFIG_SIZE => {
                if let Some(v) = data.get_mut(0) {
                    // Reading this register resets it to 0.
                    *v = self.interrupt_status.swap(0, Ordering::SeqCst) as u8;
                }
            }
...
}
```

suggestion change: cloud-hypervisor could notify guest kernel by writing interrupt_status directly for control queue handling.

```
    let cur_val = interrupt_status.load(Ordering::SeqCst);
    interrupt_status.compare_and_swap(cur_val, status as usize, Ordering::SeqCst);
```
How do you think? If you agree, I will update the PR with the suggestion change above.

**Comment by sboeuf** [comment]
@bjzhjing the code you're pointing at regarding the `ISR` is only used if the device uses `INTx`.
Here is the part of the spec that relates to our case:
```
4.1.5.3.1
Device Requirements: Used Buffer Notifications
    If MSI-X capability is enabled and queue_msix_vector is NO_VECTOR for a virtqueue,
    the device MUST NOT deliver an interrupt for that virtqueue.
```

This means we don't need to notify the driver about used buffers for the `controlq` queue.

**Comment by bjzhjing** [comment]
> @bjzhjing the code you're pointing at regarding the `ISR` is only used if the device uses `INTx`.
> Here is the part of the spec that relates to our case:
> 
> ```
> 4.1.5.3.1
> Device Requirements: Used Buffer Notifications
>     If MSI-X capability is enabled and queue_msix_vector is NO_VECTOR for a virtqueue,
>     the device MUST NOT deliver an interrupt for that virtqueue.
> ```
> 
> This means we don't need to notify the driver about used buffers for the `controlq` queue.

The spec never mentioned that virtio-net driver does not assign a valid MSIX vector for controlq, but guest kernel do it. Per my understanding, the statement here indicates that the device should not deliver a MSIX interrupt. Look at the following send command for control queue in virtio-net, you could see that it spins for a response, and will hang if not get notified by VMM. 
```
/*
 * Send command via the control virtqueue and check status.  Commands
 * supported by the hypervisor, as indicated by feature bits, should
 * never fail unless improperly formatted.
 */
static bool virtnet_send_command(struct virtnet_info *vi, u8 class, u8 cmd,
				 struct scatterlist *out)
{
	struct scatterlist *sgs[4], hdr, stat;
	unsigned out_num = 0, tmp;

	/* Caller should know better */
	BUG_ON(!virtio_has_feature(vi->vdev, VIRTIO_NET_F_CTRL_VQ));

...
	virtqueue_add_sgs(vi->cvq, sgs, out_num, 1, vi, GFP_ATOMIC);

	if (unlikely(!virtqueue_kick(vi->cvq)))
		return vi->ctrl->status == VIRTIO_NET_OK;

	/* Spin for a response, the kick causes an ioport write, trapping
	 * into the hypervisor, so the request should be handled immediately.
	 */
	while (!virtqueue_get_buf(vi->cvq, &tmp) &&
	       !virtqueue_is_broken(vi->cvq))
		cpu_relax();

	return vi->ctrl->status == VIRTIO_NET_OK;
}
```

**Comment by sboeuf** [comment]
@bjzhjing 
No I disagree, the spec is very clear. In our case, the device is using `MSI-X` and `MSI-X` is enabled, but the driver decided to set `NO_VECTOR`, which means the device should not try to send any interrupt to the guest. The spec never mentions that because of `NO_VECTOR`, the device should fallback onto legacy interrupt (`INTx`).

The code snippet you provided is actually perfect to prove the point. The function `virtqueue_kick()` will notify the hypervisor (through `ioventfd` and VM exit), this means at this point our VMM will take care of the available buffers on the control queue. Once the VMM is done reading, it will write to them and mark them as used buffers.
When the guest execution is resumed, the `virtqueue_kick()` function returns `true` because the kick succeeded, and moves on to the next function `virtqueue_get_buf()`. This function, because it knows the previous one generated a VM exit and that the VMM had the time to fill the used buffers, will read from the buffers directly. That's how it gets the response from the `controlq`.

```
        if (unlikely(!virtqueue_kick(vi->cvq)))
		return vi->ctrl->status == VIRTIO_NET_OK;

	/* Spin for a response, the kick causes an ioport write, trapping
	 * into the hypervisor, so the request should be handled immediately.
	 */
	while (!virtqueue_get_buf(vi->cvq, &tmp) &&
	       !virtqueue_is_broken(vi->cvq))
		cpu_relax();
```

**Comment by bjzhjing** [comment]
@sboeuf so, what's the conclusion? send notification regarding control queue handling done or not?

**Comment by sboeuf** [comment]
@bjzhjing 
>so, what's the conclusion? send notification regarding control queue handling done or not?

IMO, the only thing we need to do is to patch the code [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vm-virtio/src/transport/pci_device.rs#L500) to make sure we ignore the interrupt when `q.vector == NO_VECTOR`.

**Comment by bjzhjing** [comment]
@sboeuf Verify in PR #409  that not deliver notification for control queue, it works. Add the checking vector == VIRTIO_MSI_NO_VECTOR there also. This PR could be closed.

**Comment by sboeuf** [comment]
@bjzhjing 
>@sboeuf Verify in PR #409 that not deliver notification for control queue, it works. Add the checking vector == VIRTIO_MSI_NO_VECTOR there also. This PR could be closed.

Thanks :)
